src/stringToPath.test.ts

```diff
@@ -1,11 +1,68 @@
 import { stringToPath } from "./stringToPath";
 
-test("should convert a string to a deeply nested path", () => {
-  const res = stringToPath("a.b[0].c");
-  expect<["a", "b", "0", "c"]>(res).toEqual(["a", "b", "0", "c"]);
-});
+describe("stringToPath", () => {
+  describe("implementation", () => {
+    test("should convert a string to a deeply nested path", () => {
+      const res = stringToPath("foo.bar[3].baz");
+      expect(res).toEqual(["foo", "bar", "3", "baz"]);
+    });
+
+    test("should handle nested dot paths", () => {
+      const res = stringToPath("foo[bar.baz].qui.che");
+      expect(res).toEqual(["foo", "bar.baz", "qui", "che"]);
+    });
+
+    test("should handle short path with only bracket access", () => {
+      const res = stringToPath("foo[bar]");
+      expect(res).toEqual(["foo", "bar"]);
+    });
+
+    test("should handle bracket access at the end", () => {
+      const res = stringToPath("foo.bar[3]");
+      expect(res).toEqual(["foo", "bar", "3"]);
+    });
+
+    test("should handle relatively long paths without performance issues", () => {
+      const res = stringToPath(
+        "lorem.ipsum[dolor.sit].amet.con.sec.tetur[adi.pisc.ing].elit.42",
+      );
+      expect(res).toEqual([
+        "lorem",
+        "ipsum",
+        "dolor.sit",
+        "amet",
+        "con",
+        "sec",
+        "tetur",
+        "adi.pisc.ing",
+        "elit",
+        "42",
+      ]);
+    });
+  });
+
+  describe("types", () => {
+    test("simple const string are inferred", () => {
+      const input = "foo[bar.baz].qui";
+      const data = stringToPath(input);
+      assertType<["foo", "bar.baz", "qui"]>(data);
+    });
+
+    test("string constructed from const strings are inferred", () => {
+      const bar = "bar";
+      const baz = "baz";
+      const qui = "qui";
+      const input = `foo[${bar}.${baz}].${qui}.che`;
+      const data = stringToPath(input);
+      assertType<["foo", "bar.baz", "qui", "che"]>(data);
+    });
 
-test("should handle nested dot paths", () => {
-  const res = stringToPath("a.b[a.b].c");
-  expect<["a", "b", "a.b", "c"]>(res).toEqual(["a", "b", "a.b", "c"]);
+    test("dynamic strings cannot be inferred", () => {
+      // eslint-disable-next-line @typescript-eslint/no-inferrable-types
+      const bar: string = "bar";
```

Comment:
**nit:** we can skip the lint rule with a `"bar" as string` i think?

Reply:
yes, in general, `as <type>` is also safer for tests because it forces typescript to treat it as that type, and `: <type> =` actually focuses typescript more on the validity of the assignment itself; it can still widen or narrow the type based on the rest of the code.